File: java-springboot-jpa/app/src/test/java/es/msanchez/templates/java/springboot/jpa/service/PersonServiceTest.java
package es.msanchez.templates.java.springboot.jpa.service;

import es.msanchez.templates.java.springboot.jpa.dao.PersonDao;
import es.msanchez.templates.java.springboot.jpa.entity.Person;
import org.assertj.core.api.BDDAssertions;
import org.mockito.BDDMockito;
import org.mockito.Mockito;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import java.util.Optional;

public class PersonServiceTest {

    private PersonService service;

    private PersonDao daoMock;

    @BeforeMethod
    public void setUp() {
        this.daoMock = Mockito.mock(PersonDao.class);

        this.service = new PersonService(this.daoMock);
    }

    @Test
    public void testIsValidName() {
        // @GIVEN
        final String name = "mario";

        BDDMockito.given(this.daoMock.findOneByName(name))
                .willReturn(Optional.empty());

        // @WHEN
        final boolean isValid = this.service.isValid(name);

        // @THEN
        BDDMockito.verify(this.daoMock).findOneByName(name);
        BDDAssertions.assertThat(isValid).isTrue();
    }

    @Test
    public void testIsValidNameCaseFalse() {
        // @GIVEN
        final String name = "mario";

        BDDMockito.given(this.daoMock.findOneByName(name))
                .willReturn(Optional.of(new Person()));

        // @WHEN
        final boolean isValid = this.service.isValid(name);

        // @THEN
        BDDMockito.verify(this.daoMock).findOneByName(name);
        BDDAssertions.assertThat(isValid).isFalse();
    }

    @Test
    public void testSave() {
        // @GIVEN
        final Person person = new Person();
        person.setName("mario");

        // @WHEN
        this.service.save(person);

        // @THEN
        BDDMockito.verify(this.daoMock).save(person);
    }
}

File: kotlin-springboot-jpa/app/src/test/java/es/msanchez/frameworks/spring/boot/service/PersonServiceTest.kt
package es.msanchez.frameworks.spring.boot.service

import es.msanchez.frameworks.spring.boot.dao.PersonDao
import es.msanchez.frameworks.spring.boot.entity.Person
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.assertj.core.api.BDDAssertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*

@ExtendWith(MockKExtension::class)
internal class PersonServiceTest {

    @MockK
    lateinit var dao: PersonDao

    @InjectMockKs
    lateinit var service: PersonService

    @Test
    internal fun testIsValidCaseDoesntExist() {
        // Given
        val name = "mario"

        every { dao.findOneByName(name) } returns Optional.empty()

        // When
        val result = this.service.isValid(name)

        // Then
        BDDAssertions.assertThat(result).isTrue()
    }

    @Test
    internal fun testIsValidCaseAlreadyExists() {
        // Given
        val name = "mario"

        every { dao.findOneByName(name) } returns Optional.of(Person())

        // When
        val result = this.service.isValid(name)

        // Then
        BDDAssertions.assertThat(result).isFalse()
    }

    @Test
    internal fun testSave() {
        // Given
        val person = Person()

        every { dao.save(person) } returns null

        // When
        this.service.save(person)

        // Then
        verify { dao.save(person) }
    }

}